{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "f58ff9d0-8c9b-455b-9a1d-aacde7cc948f",
   "metadata": {},
   "source": [
    "# General settings"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "88d010b5-d237-43fd-b85a-ae8c007f5c8b",
   "metadata": {},
   "source": [
    "Import anything that can be useful"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "64d08973-c3ea-4866-849f-c560503445b0",
   "metadata": {},
   "outputs": [],
   "source": [
    "from sklearn.model_selection import train_test_split\n",
    "from sklearn.metrics import mean_squared_error\n",
    "from sklearn.linear_model import LinearRegression\n",
    "import random\n",
    "import bisect\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import time\n",
    "import matplotlib.pyplot as plt"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0fdba255-896a-4f9f-a814-941dca6b62a4",
   "metadata": {},
   "source": [
    "We read the dataset and fix reasonable parameters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "4c93d833-0fb1-4302-9eab-5cf924aa4092",
   "metadata": {},
   "outputs": [],
   "source": [
    "data = pd.read_excel('data.xlsx')\n",
    "data = data.dropna(axis=0)\n",
    "X = data.iloc[:, :-1]\n",
    "y = data.iloc[:, -1]\n",
    "GENES = X.shape[1]\n",
    "\n",
    "# Binary genome\n",
    "bases = [False, True]\n",
    "MUTATION_RATE = 0.1\n",
    "CROSSOVER_RATE = 0.7\n",
    "\n",
    "GENERATIONS = 20\n",
    "POPULATION_SIZE = 100\n",
    "SEED = 1"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9067a917-43a3-4fba-978b-7dd1da4ed77b",
   "metadata": {},
   "source": [
    "Implement statistic functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "673ddfb2-56ce-4f11-86bd-a7c8439e1214",
   "metadata": {},
   "outputs": [],
   "source": [
    "def mask_X(X, mask):\n",
    "    return X.iloc[:, mask]\n",
    "\n",
    "\n",
    "def random_mask(n):\n",
    "    return np.random.binomial(1, 0.5, n).astype(bool)\n",
    "\n",
    "\n",
    "def linreg(M, v):\n",
    "    X_train, X_test, y_train, y_test = train_test_split(M, v, test_size=0.2, random_state=0)\n",
    "    lr = LinearRegression()\n",
    "    lr.fit(X_train, y_train)\n",
    "    y_ = lr.predict(X_test)\n",
    "    return mean_squared_error(y_test, y_)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bd4ece7c-6ad4-450d-a407-e6043d9add23",
   "metadata": {},
   "source": [
    "# Populations"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ba31fe81-227a-45f7-9a23-e5b039241c6b",
   "metadata": {},
   "source": [
    "We define the individual object. When an individual is created their genome is generated randomly up to the number of genes needed, but the fitness is not calculated to avoid unnecesary computations. The fitness is only computed (and stored) when comparing individuals.\n",
    "\n",
    "We have also implemented multiple sources of variability. We have punctual mutations (change individual bits, or between any bases out of a set if there is more than 2 options) and transpositions (one substring gets moved to a different position) for one individual, and crossovers between 2 individuals to give 2 new individuals. Transpositions are not used since they are too destructive."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "910c8e84-8cd3-4e20-8bd4-6db7943cf0ed",
   "metadata": {},
   "outputs": [],
   "source": [
    "class Pindividual:\n",
    "    def __init__(self, chromosome):\n",
    "        length = len(list(chromosome))\n",
    "        if length > GENES:\n",
    "            # print(\"WARNING: truncating chromosome\")\n",
    "            ch = list(chromosome)[:GENES]\n",
    "        elif length < GENES:\n",
    "            # print(\"WARNING: chromosome too short\")\n",
    "            ch = list(chromosome) + [random.randint(0, 1) for i in range(GENES - length)]\n",
    "        else:\n",
    "            ch = list(chromosome)\n",
    "        self.chromosome = [bool(i) for i in ch]\n",
    "        self.fitness = 0\n",
    "\n",
    "    def __str__(self):\n",
    "        # return f\"fitness: {self.calculate_fitness()}\"\n",
    "        string = \"\"\n",
    "        for i in self.chromosome:\n",
    "            if i:\n",
    "                string = string + \"1\"\n",
    "            else:\n",
    "                string = string + \"0\"\n",
    "        return f\"{string}, fitness: {round(self.calculate_fitness())}\"\n",
    "\n",
    "    def copy(self):\n",
    "        return Pindividual(self.chromosome)\n",
    "\n",
    "    def calculate_fitness(self):\n",
    "        self.fitness = linreg(mask_X(X, self.chromosome), y)\n",
    "        return self.fitness\n",
    "\n",
    "    def __lt__(self, other):\n",
    "        if self.fitness == 0:\n",
    "            self.calculate_fitness()\n",
    "        if other.fitness == 0:\n",
    "            other.calculate_fitness()\n",
    "\n",
    "        return self.fitness < other.fitness\n",
    "\n",
    "    def crossover(self, partner):\n",
    "        midpoint = random.randint(0, len(self.chromosome))\n",
    "        child1 = self.chromosome[:midpoint] + partner.chromosome[midpoint:]\n",
    "        child2 = self.chromosome[midpoint:] + partner.chromosome[:midpoint]\n",
    "        return Pindividual(child1), Pindividual(child2)\n",
    "\n",
    "    def punctual_mutation(self, prob):\n",
    "        for i in range(len(self.chromosome)):\n",
    "            if random.random() < prob:\n",
    "                # mutable_bases = bases[:]\n",
    "                # mutable_bases.remove(self.chromosome[i])\n",
    "                # self.chromosome[i] = random.choice(mutable_bases)\n",
    "                self.chromosome[i] = not self.chromosome[i]\n",
    "        self.fitness = 0\n",
    "\n",
    "    def transposition_mutation(self, prob):\n",
    "        if random.random() > prob:\n",
    "            return\n",
    "        MAXSIZE = np.floor(GENES / 4)\n",
    "        size = random.randint(1, MAXSIZE)\n",
    "        place1 = random.randint(0, GENES - size - 1)\n",
    "        place2 = place1\n",
    "        while (abs(place1 - place2) < size):\n",
    "            place2 = random.randint(0, GENES - size - 1)\n",
    "        if place1 > place2:\n",
    "            aux = place2\n",
    "            place2 = place1\n",
    "            place1 = aux\n",
    "        print(size, place1, place2)\n",
    "        interposon1 = self.chromosome[place1:place1 + size]\n",
    "        interposon2 = self.chromosome[place2:place2 + size]\n",
    "        self.chromosome = self.chromosome[:place1] + interposon2 + self.chromosome[place1 + size:place2] + interposon1 + self.chromosome[place2 + size:]\n",
    "        self.fitness = 0"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b8bdd500-046e-43a8-ad28-d1c59c8b3a22",
   "metadata": {},
   "source": [
    "Another important function we implemented is _evolve_. It runs a selection method on a population a given amount of times and records the best individual in each generation."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "1f05d291-ff12-4c03-b799-3488bf095cf7",
   "metadata": {},
   "outputs": [],
   "source": [
    "def evolve(popu, gen, selection, param, best):\n",
    "    for i in range(gen):\n",
    "        print(f\"gen {i+1}, time: \", end='')\n",
    "        t1 = time.time()\n",
    "        popu = selection(popu, param)\n",
    "        t2 = time.time()\n",
    "        print(f\"{round(t2-t1,5)}s\")\n",
    "        # best.append(min(popu))\n",
    "    print(min(popu))\n",
    "    return popu"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "44519b63-9520-4e3c-ae62-31d18e9512de",
   "metadata": {},
   "source": [
    "# Selection Methods"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "19856afe-9917-4be5-b4da-e32f55a462b7",
   "metadata": {},
   "source": [
    "All our methods take a population and 1 parameter, and apply some technique to generate the next generation of the population. We have designed them as such to control everything from the _evolve_ function."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "17ecc621-10bb-4a5f-83ef-bf15b30152b8",
   "metadata": {},
   "source": [
    "Direct selection takes the _n_ fittest individuals of a population and saves them (as in Noah's ark) to then repopulate from them into a whole new population. The repopulation is done first with crossovers between individuals until we reach a given threshold, and then mutating the saved individuals (fittest from previous generation and new ones) until we reach the wanted population size.\n",
    "\n",
    "It usually results in the best outcomes, although it's slower since all individuals are evaluated."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "98ea3495-dbe3-4724-8a20-d3b0f8e3a7b5",
   "metadata": {},
   "outputs": [],
   "source": [
    "def direct_selection(population, ark_capacity):\n",
    "    # take best\n",
    "    saved = []\n",
    "    for i in population:\n",
    "        if len(saved) < ark_capacity:\n",
    "            bisect.insort(saved, i)\n",
    "            continue\n",
    "        if i < saved[-1]:\n",
    "            bisect.insort(saved, i)\n",
    "        if len(saved) > ark_capacity:\n",
    "            saved.pop()\n",
    "\n",
    "    # keep saved\n",
    "    new_popu = []\n",
    "    for i in saved:\n",
    "        elem = i.copy()\n",
    "        new_popu.append(elem)\n",
    "\n",
    "    # aber cojan\n",
    "    crossings = int((POPULATION_SIZE - ark_capacity) * CROSSOVER_RATE / 2)\n",
    "    for i in range(crossings):\n",
    "        # from OG (exploitation)\n",
    "        # parent1 = random.randint(0, ark_capacity - 1)\n",
    "        # parent2 = random.randint(0, ark_capacity - 1)\n",
    "        # child1, child2 = saved[parent1].crossover(saved[parent2])\n",
    "        # from new (exploration)\n",
    "        parent1 = random.randint(0, len(new_popu) - 1)\n",
    "        parent2 = random.randint(0, len(new_popu) - 1)\n",
    "        child1, child2 = new_popu[parent1].crossover(new_popu[parent2])\n",
    "\n",
    "        new_popu.append(child1)\n",
    "        new_popu.append(child2)\n",
    "\n",
    "    # fill with mutants\n",
    "    while len(new_popu) < POPULATION_SIZE:\n",
    "        pos = random.randint(0, len(new_popu) - 1)\n",
    "        elem = new_popu[pos].copy()\n",
    "        elem.punctual_mutation(MUTATION_RATE)\n",
    "        new_popu.append(elem)\n",
    "\n",
    "    return new_popu"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d55b5225-36da-48f4-b187-af5ce7e85bb8",
   "metadata": {},
   "source": [
    "Tournament selection takes a sample of the population and selects the fittest. Then, among the fittest ones we crossover until we get a new population. Some mutations are done before inserting in next generation.\n",
    "\n",
    "Since crossover is expensive, we only do it in proportion the fixed _CROSSOVER\\_RATE_.\n",
    "\n",
    "This method is faster since we do less evaluations on average, but usually doesn't reach the same results as direct selection"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "18a22ff5-c7ba-486d-832a-6fc3eba38225",
   "metadata": {},
   "outputs": [],
   "source": [
    "def tournament(population, selection_size):\n",
    "    new_popu = []\n",
    "    while len(new_popu) < POPULATION_SIZE:\n",
    "        parent1 = min(random.sample(population, selection_size))\n",
    "        parent2 = min(random.sample(population, selection_size))\n",
    "        if (random.random() < CROSSOVER_RATE):\n",
    "            ch1, ch2 = parent1.crossover(parent2)\n",
    "            ch1.punctual_mutation(MUTATION_RATE)\n",
    "            ch2.punctual_mutation(MUTATION_RATE)\n",
    "            new_popu.append(ch1)\n",
    "            new_popu.append(ch2)\n",
    "        else:\n",
    "            parent1.punctual_mutation(MUTATION_RATE)\n",
    "            parent1.fitness = 0\n",
    "            parent2.punctual_mutation(MUTATION_RATE)\n",
    "            parent2.fitness = 0\n",
    "            new_popu.append(parent1)\n",
    "            new_popu.append(parent2)\n",
    "    return new_popu"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6fb45020-83c4-4cff-b922-067b5991ba5c",
   "metadata": {},
   "source": [
    "# Analysis"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b917680d-fa54-407b-b2da-ccff6cda6e3f",
   "metadata": {},
   "source": [
    "Create 2 identical populations and use a different selection in each.\n",
    "\n",
    "Since we don't want to give any of them a time advantage, we don't evaluate the populations to get the fittest at the cost of not having plots for fitness (for now)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "b799d641-b5f6-4c89-b406-553df3a5bbfd",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "gen 0, time: 5.87915s\n",
      "gen 1, time: 4.00186s\n",
      "gen 2, time: 5.04862s\n",
      "gen 3, time: 4.69017s\n",
      "gen 4, time: 5.86879s\n",
      "gen 5, time: 6.14513s\n",
      "gen 6, time: 4.74571s\n",
      "gen 7, time: 6.13961s\n",
      "gen 8, time: 4.58077s\n",
      "gen 9, time: 6.05976s\n",
      "gen 10, time: 7.24779s\n",
      "gen 11, time: 6.71922s\n",
      "gen 12, time: 6.9208s\n",
      "gen 13, time: 6.68885s\n",
      "gen 14, time: 5.52632s\n",
      "gen 15, time: 5.12792s\n",
      "gen 16, time: 6.3786s\n",
      "gen 17, time: 5.83658s\n",
      "gen 18, time: 4.76682s\n",
      "gen 19, time: 6.50933s\n",
      "1011110011001010001010110111100101100101011, fitness: 4637022674\n",
      "Direct selection total time: 119.1733615398407s\n",
      "gen 0, time: 4.26063s\n",
      "gen 1, time: 6.0336s\n",
      "gen 2, time: 6.26672s\n",
      "gen 3, time: 6.43974s\n",
      "gen 4, time: 5.53676s\n",
      "gen 5, time: 5.73761s\n",
      "gen 6, time: 6.1825s\n",
      "gen 7, time: 5.84574s\n",
      "gen 8, time: 7.03976s\n",
      "gen 9, time: 7.15841s\n",
      "gen 10, time: 6.81911s\n",
      "gen 11, time: 6.64393s\n",
      "gen 12, time: 5.65876s\n",
      "gen 13, time: 6.9498s\n",
      "gen 14, time: 5.70735s\n",
      "gen 15, time: 4.96238s\n",
      "gen 16, time: 5.70612s\n",
      "gen 17, time: 5.90016s\n",
      "gen 18, time: 6.95658s\n",
      "gen 19, time: 4.64134s\n",
      "1111010000001110001010111110111010100001011, fitness: 4631489155\n",
      "Direct selection total time: 124.21885323524475s\n"
     ]
    }
   ],
   "source": [
    "random.seed(SEED)\n",
    "popu1 = [Pindividual([]) for _ in range(POPULATION_SIZE)]\n",
    "popu2 = []\n",
    "for i in popu1:\n",
    "    popu2.append(i.copy())\n",
    "\n",
    "t1 = time.time()\n",
    "evolved = evolve(popu1, GENERATIONS, tournament, 10,[])\n",
    "t2 = time.time()\n",
    "print(f\"Direct selection total time: {t2-t1}s\")\n",
    "\n",
    "t1 = time.time()\n",
    "evolved = evolve(popu2, GENERATIONS, direct_selection, int(POPULATION_SIZE / 10),[])\n",
    "t2 = time.time()\n",
    "print(f\"Direct selection total time: {t2-t1}s\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
